src/controllers/aer_controller.hpp

```diff
@@ -1404,85 +1525,262 @@ Transpile::Fusion Controller::transpile_fusion(Method method,
+  catch (std::exception &e) {
+    result.status = ExperimentResult::Status::error;
+    result.message = e.what();
   }
+}
 
-  // Run multishot simulation without noise sampling
-  run_circuit_without_sampled_noise(opt_circ, config, shots, state,
-                                    method, cache_blocking, result, rng);
-  state.add_metadata(result);
+template <class State_t>
+void Controller::run_batched_circuits_helper(const std::vector<Circuit> &circs,
+                                    const Noise::NoiseModel &noise,
+                                    const json_t &config,
+                                    const Method method,
+                                    Result &result) const
+{
+  auto timer_start = myclock_t::now(); // state circuit timer
+
+#pragma omp parallel for
+  for (int j = 0; j < circs.size(); ++j) {
+    // Initialize circuit json return
+    result.results[j].legacy_data.set_config(config);
+
+    // Output data container
+    result.results[j].set_config(config);
+    result.results[j].metadata.add(method_names_.at(method), "method");
+    if (method == Method::statevector || method == Method::density_matrix ||
+        method == Method::unitary) {
+      result.results[j].metadata.add(sim_device_name_, "device");
+    } else {
+      result.results[j].metadata.add("CPU", "device");
+    }
+    // Add measure sampling to metadata
+    // Note: this will set to `true` if sampling is enabled for the circuit
+    result.results[j].metadata.add(false, "measure_sampling");
+
+    // Circuit qubit metadata
+    result.results[j].metadata.add(circs[j].num_qubits, "num_qubits");
+    result.results[j].metadata.add(circs[j].num_memory, "num_clbits");
+    result.results[j].metadata.add(circs[j].qubits(), "active_input_qubits");
+    result.results[j].metadata.add(circs[j].qubit_map(), "input_qubit_map");
+    result.results[j].metadata.add(circs[j].remapped_qubits, "remapped_qubits");
+  }
+
+  // Execute in try block so we can catch errors and return the error message
+  // for individual circuit failures.
+  try {
+    State_t state;
+    Multi::States<State_t> states;
+
+    int_t i_circ;
+
+    std::vector<RngEngine> rng(circs.size());
+    std::vector<std::vector<Operations::Op>> ops(circs.size());
+    std::vector<std::vector<Operations::Op>> meas_roerror_ops(circs.size());
+    std::vector<double> global_phase(circs.size());
+
+    int_t max_bits = 1;
+    for (i_circ=0;i_circ< circs.size(); i_circ++) {
+      max_bits = std::max(max_bits,get_max_matrix_bits(circs[i_circ]) );
+    }
+    states.set_max_matrix_bits(max_bits);
+
+    states.set_parallelization(max_batched_states_);
+    states.allocate(max_qubits_, max_qubits_,circs.size());
+    states.set_config(config);
+
+#pragma omp parallel for if(circs.size() > 1)
+    for (i_circ=0;i_circ< circs.size(); i_circ++) {
+      rng[i_circ].set_seed(circs[i_circ].seed);
+
+      Circuit circ;
```

Comment:
We can reduce duplication to generate a sampling circuit.

Reply:
created a function for noise sampling